src/lib.rs

```diff
@@ -228,18 +227,12 @@ where
         ctx: &mut Context<'_>,
         buf: &mut ReadBuf<'_>,
     ) -> Poll<io::Result<()>> {
-        self.with_context(ctx, |s| {
-            // This isn't really "proper", but rust-openssl doesn't currently expose a suitable interface even though
-            // OpenSSL itself doesn't require the buffer to be initialized. So this is good enough for now.
-            let slice = unsafe {
-                let buf = buf.unfilled_mut();
-                slice::from_raw_parts_mut(buf.as_mut_ptr().cast::<u8>(), buf.len())
-            };
-            match cvt(s.read(slice))? {
+        // SAFETY: read_uninit does not de-initialize the buffer and guarantees that the first nread
+        // bytes are initialized.
+        self.with_context(ctx, |s| unsafe {
+            match cvt(s.read_uninit(buf.unfilled_mut()))? {
                 Poll::Ready(nread) => {
-                    unsafe {
-                        buf.assume_init(nread);
-                    }
+                    buf.assume_init(nread);
                     buf.advance(nread);
                     Poll::Ready(Ok(()))
```

Comment:
Generally you should avoid so large unsafe blocks.
```rs
// SAFETY: The `read_uninit` method promises to never deinitialize any of the bytes.
match cvt(s.read_uninit(unsafe { buf.unfilled_mut() }))? {
    Poll::Ready(nread) => {
        // SAFETY: The `read_uninit` method promises that the first `nread` bytes of the buffer are now initialized.
        unsafe { buf.assume_init(nread) };

Reply:
Updated! Sorry for the delay.